Changelog tool: Enhance entry normalizations

aduth opened the pull request:
This pull request seeks to enhance the plugin changelog generator tool to further enhance normalizations applied to release note entries:

- Omit entries by label (e.g. "Mobile App Compatibility", "Project Management")
- Omit by title prefix (e.g. "[RNMobile]")
- Rename headings (e.g. "Bug" → "Bug Fixes")
- Group related headings (e.g. "Bug" and "Regression" → "Bug Fixes")
- Order headings (e.g. "Features" at the top)
- Reword common technical terms (e.g. "e2e" → "end-to-end")
- Capitalize first meaningful segment (e.g. "e2e-tests: fix bug" → "e2e-tests: Fix bug")
   - Note: I considered to try to remove these prefixing, but I opted against it, since they can be useful for manual curation of entries, to reorganize entries to appropriate headings. This could be further enhanced in the future.

**Edit:** More enhancements added 2020-05-26:

- Derive type from pull request title (e.g. "Fixes ..." -> "Bug Fixes")
- Prioritizes assigned groups (e.g. "Enhancement" over "Task"/"Various")
- Remove redundant prefixes from title (e.g. remove prefix from "Code Quality: ..." if group is already known as "Code Quality")

All of the above were implemented in a way to facilitate future enhancements, e.g. to make it easy to omit more entries by label name.

**Testing Instructions:**

Ensure unit tests pass:

```
npm run test-unit bin/plugin/commands/test/changelog.js
```

Generate a changelog to assess the output is sensible:

```
npm run changelog
```

aduth committed:
Build Tools: Refactor entry normalization as flow

aduth committed:
Build Tools: Changelog: Omit mobile entries

aduth committed:
Build Tools: Reword common technical terms

aduth committed:
Build Tools: Changelog: Capitalize segment after colon

aduth committed:
Build Tools: Changelog: Remap various label group headings

aduth committed:
Build Tools: Changelog: Sort group titles

aduth committed:
Build Tools: Omit project management label

youknowriad commented on bin/plugin/commands/changelog.js:
```diff
@@ -33,6 +33,47 @@ const manifest = require( '../../../package.json' );
+	'Code Quality',
+	'Documentation',
+	'Project Management',
+	undefined,
```
What is undefined here?

aduth replied:
> What is undefined here?

It's a little opaque, but explained in the comment:

https://github.com/WordPress/gutenberg/blob/a24fe1117b5bc680ea1d686fb46a5efb7fd467de/bin/plugin/commands/changelog.js#L59-L60

youknowriad replied:
Do you think we can use the "grouping" like here https://github.com/WordPress/gutenberg/pull/22601/files#diff-ea435cc89a172cc49a2ab3b0ff264cc1R52 to automatically move these "undefined" items under "Various" instead?

aduth replied:
The thinking is to try to avoid having to have built-in awareness of every possible heading, while at the same time wanting "Various" as both known and intending to be the very-last heading.

youknowriad replied:
Oh, it's more clear now, thanks :)

youknowriad commented on bin/plugin/commands/changelog.js:
```diff
@@ -33,6 +33,47 @@ const manifest = require( '../../../package.json' );
+	Feature: 'Features',
+	Enhancement: 'Enhancements',
+	'New API': 'New APIs',
+};
```
I think we should add "Experimental": "Experiments" here right?

aduth replied:
> I think we should add "Experimental": "Experiments" here right?

Yeah. Updated in 064a3dd.

Alternatively, the label "[Type] Experimental" could be renamed "[Type] Experiments". For future consideration.

youknowriad commented on bin/plugin/commands/changelog.js:
```diff
@@ -33,6 +33,47 @@ const manifest = require( '../../../package.json' );
+	'New API': 'New APIs',
+};
+
+/**
```
I noticed that I had a "Task" heading when running the script today, I wonder if we should remove this generic label to force us to choose something more meaningful (feature, enhancement...)

aduth replied:
> I noticed that I had a "Task" heading when running the script today, I wonder if we should remove this generic label to force us to choose something more meaningful (feature, enhancement...)

Interesting question. It does seem like "Task" is used as something of a fallback, when "Bug" or "Enhancement" don't quite fit. It may have been inherited from how Types are assigned in core in [ticket creation](https://core.trac.wordpress.org/newticket) ("defect (bug)", "enhancement", "feature request", "task (blessed)").

In the meantime, 22d8fce remaps "Task" to "Various", so at least it's lumped in with the rest of the miscellany.

youknowriad reviewed (approved):
Used it today and it was great.

aduth committed:
Build Tools: Changelog: Remap Experimental as Experiments

aduth committed:
Build Tools: Changelog: Move reword constant to top

Ideally all constants defined at top, but TITLE_NORMALIZATIONS depends on functions defined before it's declared. Alternatively it could be inlined into the normalization function, rather than defined as a constant.

aduth committed:
Build Tools: Changelog: Map Task type to Various

aduth committed:
Build Tools: Changelog: Order Code Quality before Documentation

aduth committed:
Build Tools: Changelog: Prioritize chosen type by group order

aduth committed:
Build Tools: Changelog: Match type by title pattern

aduth committed:
Build Tools: Changelog: Correctly omit by project management

aduth committed:
Build Tools: Changelog: Remove redundant type prefixes

aduth committed:
Build Tools: Changelog: Fix to omit empty groups

aduth committed:
Build Tools: Changelog: Add performance section to ordered groups

aduth committed:
Build Tools: Changelog: Add additional rewording (config, docs)

aduth committed:
Build Tools: Changelog: Trim trailing whitespace before appending period

aduth committed:
Build Tools: Changelog: Update sortGroup test for ordered Performance

fixup 59dd9a6827

youknowriad commented:
Awesome work, thanks for the last improvements.

aduth commented:
I updated the original comment with the additional enhancements I implemented yesterday:

>**Edit:** More enhancements added 2020-05-26:
>
>- Derive type from pull request title (e.g. "Fixes ..." -> "Bug Fixes")
>- Prioritizes assigned groups (e.g. "Enhancement" over "Task"/"Various")
>- Remove redundant prefixes from title (e.g. remove prefix from "Code Quality: ..." if group is already known as "Code Quality")
